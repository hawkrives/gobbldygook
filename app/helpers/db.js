'use strict';

import * as Promise from 'bluebird'
import * as treo from 'treo'
import * as treoPromise from 'treo/plugins/treo-promise'

var schema = treo.schema()
	.version(1)
		.addStore('courses', { key: 'clbid' })
			.addIndex('profs',      'profs',      {multi: true})
			.addIndex('depts',      'depts',      {multi: true})
			.addIndex('places',     'places',     {multi: true})
			.addIndex('times',      'times',      {multi: true})
			.addIndex('gereqs',     'gereqs',     {multi: true})
			.addIndex('sourcePath', 'sourcePath', {multi: false})
			.addIndex('crsid',      'crsid',      {multi: false})
			.addIndex('clbid',      'clbid',      {multi: false})
			.addIndex('deptnum',    'deptnum',    {multi: false})
		.addStore('areas', { key: 'sourcePath' })
			.addIndex('type',       'type',       {multi: true})
			.addIndex('sourcePath', 'sourcePath', {multi: false})
	.version(2)
		.addStore('students', { key: 'id'   })

var db = treo('gobbldygook', schema)
	.use(treoPromise())

window.eraseDatabase = function() {
	window.database.drop().then(function() {
		console.log('Database dropped')
		localStorage.clear()
		console.log('localStorage cleared')
	})
}

window.database = db
export default db
