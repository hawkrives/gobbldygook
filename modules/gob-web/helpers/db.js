// @flow

import {createDatabase} from '@gob/web-database'

export const db = createDatabase()
