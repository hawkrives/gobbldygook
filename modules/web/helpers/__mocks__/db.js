/* eslint-env jest */
// @flow

// $FlowFixMe
import setGlobals from 'indexeddbshim'
let dfn = Object.defineProperty
// $FlowFixMe
Object.defineProperty = null
global.window = global
setGlobals(global.window)
// $FlowFixMe
Object.defineProperty = dfn

// $FlowFixMe
const { default: db } = require.requireActual('../db')

db.__clear = function clearDatabase() {
    return Promise.all(db.stores.map(s => db.store(s).clear()))
}

export default db
