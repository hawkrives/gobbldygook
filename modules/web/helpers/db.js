import treo, { Database } from 'treo'
treo.Promise = global.Promise
import debug from 'debug'
const log = debug('web:database')

import queryTreoDatabase from '../helpers/treo-plugins/query-treo-database'
import batchGet from '../helpers/treo-plugins/treo-batch-get'

import schema from './db-schema'

const db = new Database('gobbldygook', schema)
	.use(queryTreoDatabase())
	.use(batchGet())

export default db


if (typeof window !== 'undefined') {
	window.deleteDatabase = () => {
		const DBDeleteRequest = window.indexedDB.deleteDatabase('gobbldygook')
		log('Commencing database deletion')
		DBDeleteRequest.onerror = () => log('Error deleting database.')
		DBDeleteRequest.onsuccess = () => log('Database deleted successfully')
	}

	window.eraseStorage = () => {
		log('Commencing storage erasure')
		window.localStorage.clear()
		log('Storage erased')
	}

	window.eraseDatabase = () => {
		window.deleteDatabase()
		window.eraseStorage()
	}

	window.database = db
}
