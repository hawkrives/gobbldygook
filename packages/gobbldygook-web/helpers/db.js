import treo, {Database} from 'treo'
treo.Promise = require('bluebird')

import queryTreoDatabase from '../helpers/treo-plugins/query-treo-database'
import batchGet from '../helpers/treo-plugins/treo-batch-get'

import schema from './db-schema'

const db = new Database('gobbldygook', schema)
	.use(queryTreoDatabase())
	.use(batchGet())

export default db


if (typeof window !== 'undefined') {
	window.deleteDatabase = () => {
		const DBDeleteRequest = window.indexedDB.deleteDatabase('gobbldygook')
		console.log('Commencing database deletion')
		DBDeleteRequest.onerror = () => console.log('Error deleting database.')
		DBDeleteRequest.onsuccess = () => console.log('Database deleted successfully')
	}

	window.eraseStorage = () => {
		console.log('Commencing storage erasure')
		window.localStorage.clear()
		console.log('Storage erased')
	}

	window.eraseDatabase = () => {
		window.deleteDatabase()
		window.eraseStorage()
	}

	window.database = db
}
