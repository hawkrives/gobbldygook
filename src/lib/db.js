// set up PouchDB
import PouchDB from 'pouchdb'
import find from 'pouchdb-find'
PouchDB.plugin(find)
import search from 'pouchdb-quick-search'
PouchDB.plugin(search)

// if we're running in node, make a directory for the databases
const prefix = process.env.NODE_ENV === 'cli' ? './databases/' : ''
// if (process.env.NODE_ENV === 'cli') {
// 	const mkdir = require('mkdirp')
// 	mkdir.sync(prefix)
// }

let opts = {}
// if we're running a test, just stick the database in memory
if (process.env.NODE_ENV === 'test') {
	opts.db = require('memdown')
}


// make the databases
export const courseDb  = new PouchDB(prefix + 'courses',  {...opts, size: 50})
export const areaDb    = new PouchDB(prefix + 'areas',    {...opts, size: 5})
export const studentDb = new PouchDB(prefix + 'students', {...opts, size: 5})
export const cacheDb   = new PouchDB(prefix + 'cache',    {...opts, size: 1})
global.database = {courseDb, areaDb, studentDb, cacheDb}

// set up course indices
courseDb.createIndex({index: {fields: ['clbid']}})  // should be unique
courseDb.createIndex({index: {fields: ['credits']}})
courseDb.createIndex({index: {fields: ['crsid']}})
courseDb.createIndex({index: {fields: ['dept']}})
courseDb.createIndex({index: {fields: ['depts']}})  // is an array
courseDb.createIndex({index: {fields: ['deptnum']}})
courseDb.createIndex({index: {fields: ['desc']}})
courseDb.createIndex({index: {fields: ['gereqs']}})  // is an array
courseDb.createIndex({index: {fields: ['groupid']}})
courseDb.createIndex({index: {fields: ['grouptype']}})
courseDb.createIndex({index: {fields: ['level']}})
courseDb.createIndex({index: {fields: ['name']}})
courseDb.createIndex({index: {fields: ['notes']}})
courseDb.createIndex({index: {fields: ['num']}})
courseDb.createIndex({index: {fields: ['pf']}})
courseDb.createIndex({index: {fields: ['locations']}})  // is an array
courseDb.createIndex({index: {fields: ['instructors']}})  // is an array
courseDb.createIndex({index: {fields: ['section']}})
courseDb.createIndex({index: {fields: ['semester']}})
courseDb.createIndex({index: {fields: ['term']}})
courseDb.createIndex({index: {fields: ['times']}})  // is an array
courseDb.createIndex({index: {fields: ['title']}})
courseDb.createIndex({index: {fields: ['type']}})
courseDb.createIndex({index: {fields: ['year']}})
courseDb.createIndex({index: {fields: ['sourcePath']}})
courseDb.createIndex({index: {fields: ['words']}})  // is an array
courseDb.createIndex({index: {fields: ['profWords']}})  // is an array

// set up area indices
areaDb.createIndex({index: {fields: ['type']}})  // is an array
areaDb.createIndex({index: {fields: ['sourcePath']}})
