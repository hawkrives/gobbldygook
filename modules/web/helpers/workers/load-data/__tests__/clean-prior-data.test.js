/* eslint-env jest */
// @flow

jest.mock('../../../db')

import db from '../../../db'
import cleanPriorData, {
    getPriorCourses,
    getPriorAreas,
} from '../clean-prior-data'

describe('getPriorCourses', () => {
    beforeEach(async () => {
        await db.__clear()
    })

    test('returns the deletion batch', async () => {
        const actions = [{ sourcePath: 'dir/file', clbid: '1' }]
        await db.store('courses').batch(actions)
        expect(Object.keys(await getPriorCourses('dir/file'))).toHaveLength(1)
    })
})

describe('getPriorAreas', () => {
    beforeEach(async () => {
        await db.__clear()
    })

    test('returns the deletion batch', async () => {
        const actions = [{ sourcePath: 'dir/file' }]
        await db.store('areas').batch(actions)
        expect(Object.keys(await getPriorAreas('dir/file'))).toHaveLength(1)
    })
})

describe('cleanPriorData', () => {
    beforeEach(async () => {
        await db.__clear()
    })

    test('clears courses', async () => {
        const actions = [{ sourcePath: 'dir/file', clbid: '1' }]
        await db.store('courses').batch(actions)

        expect(await db.store('courses').getAll()).toHaveLength(1)

        await cleanPriorData('dir/file', 'courses')

        expect(await db.store('courses').getAll()).toHaveLength(0)
    })

    test('clears areas', async () => {
        const actions = [{ sourcePath: 'dir/file' }]
        await db.store('areas').batch(actions)

        expect(await db.store('areas').getAll()).toHaveLength(1)

        await cleanPriorData('dir/file', 'areas')

        expect(await db.store('areas').getAll()).toHaveLength(0)
    })

    test('throws on an unknown type', async () => {
        expect.assertions(1)

        try {
            // $FlowFixMe
            await cleanPriorData('path', 'invalid_type')
        } catch (err) {
            expect(err.message).toMatchSnapshot()
        }
    })
})
