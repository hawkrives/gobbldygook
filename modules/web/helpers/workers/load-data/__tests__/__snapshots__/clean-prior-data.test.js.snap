// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanPriorData throws on an unknown type 1`] = `"cleanPriorData: \\"invalid_type\\" is not a valid store type"`;
